### DISCUSSION TITLE
Indices of Jacobian and Hessians?

### DISCUSSION THREAD

**Discussion by NLaws** [JuliaSmoothOptimizers/NLPModels.jl#428] [Q&A] [answered]
How do I know which variables correspond with the indices of the Jacobian and Hessian (returned from methods such as `NLPModels.jac_nln`)?  Do they use the same order as returned from `JuMP.all_variables`? (Which I believe keeps the order in which the variables were declared).

**Comment by tmigot** [comment, marked as answer]
Hi [@NLaws](https://github.com/NLaws) ! Sorry for the late reply. You had the right intuition, the order of the variables is the same as JuMP.all_variables. The order of constraints might be slightly different, the linear ones appear first, and then the nonlinear ones.

We should probably add more facilities to access the variables from the NLP model version of JuMP using the names given in the model. I opened two issues following your question:
- https://github.com/JuliaSmoothOptimizers/NLPModelsJuMP.jl/issues/145
- https://github.com/JuliaSmoothOptimizers/NLPModelsJuMP.jl/issues/144

**Reply by amontoison** [reply]
I confirm that we use the same order of `JUMP.all_variables`: https://github.com/JuliaSmoothOptimizers/NLPModelsJuMP.jl/blob/main/src/utils.jl#L292
